src/Sylius/Bundle/AddressingBundle/Repository/ZoneRepository.php

```diff
@@ -0,0 +1,121 @@
+{
+    public function findOneByAddressAndType(AddressInterface $address, string $type, ?string $scope = null): ?ZoneInterface
+    {
+        $query = $this->createByAddressQueryBuilder($address, $scope);
+
+        $query
+            ->andWhere($query->expr()->eq('o.type', ':type'))
+            ->setParameter('type', $type)
+            ->setMaxResults(1)
+        ;
+
+        return $query->getQuery()->getOneOrNullResult();
+    }
+
+    /** @return ZoneInterface[] */
+    public function findByAddress(AddressInterface $address, ?string $scope = null): array
+    {
+        return $this->createByAddressQueryBuilder($address, $scope)->getQuery()->getResult();
+    }
+
+    public function createByAddressQueryBuilder(AddressInterface $address, ?string $scope = null): QueryBuilder
+    {
+        $query = $this->createQueryBuilder('o')
+            ->select('o', 'members')
+            ->leftJoin('o.members', 'members')
+        ;
+
+        if (null !== $scope) {
+            $query
+                ->andWhere($query->expr()->in('o.scope', ':scopes'))
+                ->setParameter('scopes', [$scope, Scope::ALL])
+            ;
+        }
+
+        $orConditions = [];
+
+        if ($address->getCountryCode() !== null) {
+            $orConditions[] = $query->expr()->andX(
+                $query->expr()->eq('o.type', ':country'),
+                $query->expr()->eq('members.code', ':countryCode'),
+            );
+
+            $query->setParameter('country', ZoneInterface::TYPE_COUNTRY);
+            $query->setParameter('countryCode', $address->getCountryCode());
+        }
+
+        if ($address->getProvinceCode() !== null) {
+            $orConditions[] = $query->expr()->andX(
+                $query->expr()->eq('o.type', ':province'),
+                $query->expr()->eq('members.code', ':provinceCode'),
+            );
+
+            $query->setParameter('province', ZoneInterface::TYPE_PROVINCE);
+            $query->setParameter('provinceCode', $address->getProvinceCode());
+        }
+
+        $query->andWhere($query->expr()->orX(...$orConditions));
+
+        return $query;
+    }
+
+    /**
+     * @param array<ZoneInterface> $members
+     *
+     * @return array<ZoneInterface>
+     */
+    public function findByMembers(array $members, ?string $scope = null): array
+    {
+        $zonesCodes = array_map(
+            fn (ZoneInterface $zone): string => $zone->getCode(),
+            $members,
+        );
+
+        $query = $this->createQueryBuilder('o')
+            ->select('o', 'members')
+            ->leftJoin('o.members', 'members')
+        ;
+
+        if (null !== $scope) {
+            $query
+                ->andWhere($query->expr()->in('o.scope', ':scopes'))
+                ->setParameter('scopes', array_unique([$scope, Scope::ALL]))
+            ;
+        }
+
+        $query
+            ->andWhere('o.type = :type')
+            ->andWhere($query->expr()->in('members.code', ':zones'))
```

Comment:
"Direct" syntax is more common to see in sylius repos and IMO more readable.
Doc [examples](https://www.doctrine-project.org/projects/doctrine-orm/en/2.16/reference/query-builder.html) tend to use `expr()` only for `orX`/`andX` cases.

```suggestion
            ->andWhere('members.code IN :zones')
```

Reply:
We use both across the Sylius, there's no standard, maybe that's an idea to standarize it 🤷🏼‍♂️